plugin/pulpcore/plugin/download/futures/batch.py

```diff
@@ -52,33 +46,46 @@ class Batch:
         >>>
     """
 
-    def __init__(self, downloads, concurrent=CONCURRENT, backlog=BACKLOG, context=None):
+    def __init__(self, downloads, concurrent=CONCURRENT, context=None):
         """
         Args:
             downloads (collections.abc.Iterable): An Iterable of downloads.
-            concurrent (int): The number of downloads to execute in concurrently.
-            backlog (int): The number of downloads kept in memory.
+            concurrent (int): The number of downloads to be executed concurrently.
             context (SharedContext): An (optional) shared download context.
-
-        Raises:
-            ValueError: concurrent less than 2 or backlog is less than concurrent.
         """
         super(Batch, self).__init__()
-        self.downloads = downloads
+        self.downloads = iter(downloads)
         self.concurrent = concurrent
-        self.iterator = PlanIterator(backlog)
-        self.executor = BatchExecutor(concurrent=concurrent, backlog=backlog)
+        self.iterator = PlanIterator(self)
+        self.executor = ThreadPoolExecutor(max_workers=concurrent)
         self.context = context or Context()
-        self.feeder = DownloadFeeder(self)
-        self._is_shutdown = False
+        self.outstanding = 0
 
-    @property
-    def is_shutdown(self):
+    def submit(self, plan):
         """
-        Returns:
-            bool: Batch has been shutdown.
+        Submit the plan for execution and increment `outstanding`.
+
+        Args:
+            plan (Plan): The plan to submit.
+        """
+        download = plan.download
+        download.context = self.context
+        future = self.executor.submit(plan)
+        future.add_done_callback(self.iterator.add)
+        self.outstanding += 1
+
+    def completed(self, plan):
         """
-        return self._is_shutdown
+        A plan has completed.
+
+        Args:
+            plan (Plan): The completed plan.
+        """
+        self.outstanding -= 1
+        for download in self.downloads:
```

Comment:
I don't think accessing an iterator is necessarily thread-safe, so it probably needs to acquire a lock here.

Reply:
True.  good catch.